### ISSUE TITLE
ZFS device failure is poorly handled; insufficient privileges error with status -v

### ISSUE THREAD

**Issue by kpande** [zfsonlinux/zfs#3256]
The following was observed using two iSCSI disks exported from a low-end JBOD NAS. When the device's iSCSI target died, ZoL would not allow me to clear, export, or do anything to fix the issue without a host reboot. Not only that, but I get some "insufficient privileges" error when doing status -v.

```
[:)] # zpool status -v
  pool: athena
 state: UNAVAIL
status: One or more devices are faulted in response to IO failures.
action: Make sure the affected devices are connected, then run 'zpool clear'.
   see: http://zfsonlinux.org/msg/ZFS-8000-HC
  scan: none requested
config:

        NAME                                                       STATE     READ WRITE CKSUM
        athena                                                     UNAVAIL      0     0     0  insufficient replicas
          ip-192.168.0.32:3260-iscsi-iqn.2015.athena:drive1-lun-0  FAULTED      6     0     0  too many errors

errors: List of errors unavailable (insufficient privileges)
```

edit: I said two disks above but I hadn't attached the second disk yet.

**Comment by GregorKopka** [comment]
When a single drive vdev fails pool I/O will be halted, but you should be able to regain access to the pool in case restore access to the target device.

In case you can't do that a `zpool destroy` (maybe -f) *should* work on the pool.

**Comment by ToxicFrog** [comment]
I think https://github.com/zfsonlinux/zfs/issues/4031 is a duplicate of this.

The same thing happens to my pools after a disk failure, and has on 0.6.3, 0.6.4, and 0.6.5.

**Comment by ToxicFrog** [comment]
Still happening on 0.6.5.6; "sudo zpool clear" works, but "sudo zpool status -v" returns "insufficient privileges" instead of the list of errors.

**Comment by mailinglists35** [comment]
reproductible with the below on 0.6.5.7, discovered this accidentally while trying something else:
kernel [stack.txt](https://github.com/zfsonlinux/zfs/files/321471/stack.txt)

`truncate -s 1GiB /mnt/test/deleteme`
`zpool create deleteme /mnt/test/deleteme`
(fill in some data, I have created 2000 datasets)
`truncate -s 100MiB /mnt/test/deleteme`

at this point, `zpool scrub` hangs, and the following commands get this output:

`zpool status -v`

  pool: deleteme
 state: UNAVAIL
status: One or more devices are faulted in response to IO failures.
action: Make sure the affected devices are connected, then run 'zpool clear'.
   see: http://zfsonlinux.org/msg/ZFS-8000-HC
  scan: scrub in progress since Sat Jun 18 03:10:31 2016
    13.9M scanned out of 44.3M at 32.6K/s, 0h15m to go
    0 repaired, 31.36% done
config:

        NAME                  STATE     READ WRITE CKSUM
        deleteme              UNAVAIL      0     0     0  insufficient replicas
          /mnt/test/deleteme  UNAVAIL      0     0     0  corrupted data
errors: List of errors unavailable (insufficient privileges)

`zpool clear deleteme`
cannot clear errors for deleteme: I/O error

`zpool export deleteme`
cannot export 'deleteme': pool I/O is currently suspended

`zpool destroy deleteme`
cannot open 'deleteme': pool I/O is currently suspended

`zpool get failmode deleteme`
NAME      PROPERTY  VALUE     SOURCE
deleteme  failmode  wait      default

`zpool set failmode=continue deleteme`
cannot set property for 'deleteme': pool I/O is currently suspended

`ps ax|grep zpool`
20672 pts/3    D+     0:00 zpool scrub deleteme

`zpool scrub -s deleteme`
cannot cancel scrubbing deleteme: pool I/O is currently suspended

`rmmod -f zfs`
rmmod: ERROR: ../libkmod/libkmod-module.c:777 kmod_module_remove_module() could not remove 'zfs': Device or resource busy
rmmod: ERROR: could not remove module zfs: Device or resource busy

`ps ax|grep D`
16538 ?        D      0:04 [txg_sync]
20672 pts/3    D+     0:00 zpool scrub deleteme
23297 pts/4    D+     0:00 rmmod -f zfs

**Comment by mailinglists35** [comment]
this is still occuring on 0.7.0-rc2
iscsi target went down, zpool status shows errors, zpool status -v tells insufficient permissions, zpool online and zpool clear say "cannot clear errors for pool: I/O error"
making a self reminder to never use zol with iscsi target in production...

**Comment by dweeezil** [comment]
@mailinglists35 In your failure when the iscsi target went down and then came back, did it come back with exact same device node as is referred to by the pool according to ```zpool status```?  If the pool was using stable names (from /dev/disk/by-X), a ```zpool clear``` _should_ work because it'll re-open the vdevs with the stable names and then un-suspended it allowing IO to pick up where it left off.  At least when ```failmode=wait``` this is what should happen.

**Comment by dweeezil** [comment]
For completeness, I just tested this on my test rig by "unplugging" a pair of vdevs from a 3-vdev raidz with ```echo 1 > /sys/block/device/delete``` for each one and once the pool was suspended, re-scanned with ```echo '- - -' > /sys/class/scsi_host/host2/scan```.  When the disks came back, a ```zpool clear``` started it up again and everything was fine.

**Comment by GregorKopka** [comment]
I had the same experience: As soon a pool reaches this state I am basically done and the only action left for a remote machine is a swift ```echo b > /proc/sysrq-trigger``` as a normal reboot will hang on unmounting/exporting (with sshd already being down) which would lead to the machine being unrecoverable without physical manual interaction. 

There really needs to be a way to cleanly remove a pool from the system, regardless of the state it is in.
Issue https://github.com/zfsonlinux/zfs/issues/2878 sums it up.

**Comment by mailinglists35** [comment]
> @mailinglists35 In your failure when the iscsi target went down and then came back, did it come back with exact same device node

yes, it was sdb and came back the same (sdb)

i btraced /dev/sdb (blktrace) and when i attempt to zpool clear i see some kernel threads accessing it but no zfs threads. i am away from that system for a couple of weeks and i'll be able to paste output next time

**Comment by sanjeevbagewadi** [comment]
Observed a similar issue ("zpool status"/"zpool list" get stuck in the kernel) when the pool is faulted.
Looking into the kernel stacks. Found the following :

- The zpool status thread was stuck waiting for spa->spa_errlog_lock
-- snip --
PID: 111549 TASK: ffff8802b02c3fc0 CPU: 1 COMMAND: "zpool"
#0 [ffff8802649ffbb0] __schedule at ffffffff816cb514
#1 [ffff8802649ffc60] schedule at ffffffff816cbc10
#2 [ffff8802649ffc80] schedule_preempt_disabled at ffffffff816cbe4e
#3 [ffff8802649ffc90] __mutex_lock_slowpath at ffffffff816cd440
#4 [ffff8802649ffd30] mutex_lock at ffffffff816cd4f3
#5 [ffff8802649ffd50] spa_get_errlog_size at ffffffffa0999ae3 [zfs]
#6 [ffff8802649ffda0] spa_get_stats at ffffffffa099632c [zfs]
#7 [ffff8802649ffe00] zfs_ioc_pool_stats at ffffffffa09cb7a1 [zfs]
#8 [ffff8802649ffe40] zfsdev_ioctl at ffffffffa09cda8b [zfs]
#9 [ffff8802649ffeb0] do_vfs_ioctl at ffffffff81216072
#10 [ffff8802649fff00] sys_ioctl at ffffffff81216402
#11 [ffff8802649fff50] entry_SYSCALL_64_fastpath at ffffffff816cf76e
RIP: 00007fed39acca77 RSP: 00007ffe64af99c8 RFLAGS: 00000246
RAX: ffffffffffffffda RBX: 00007fed39d7a120 RCX: 00007fed39acca77
RDX: 00007ffe64af9a00 RSI: 0000000000005a05 RDI: 0000000000000003
RBP: 0000000000bb4e00 R8: 0000000000000008 R9: 0000000000200000
R10: 0000000000000000 R11: 0000000000000246 R12: 000000000000b210
R13: 0000000000bb4df0 R14: 0000000000010000 R15: 00007fed39d7a120
ORIG_RAX: 0000000000000010 CS: 0033 SS: 002b
-- snip --

And the spa->spa_errlog_lock is held by spa_sync() thread which is blocked on IO :
– snip –
PID: 24577 TASK: ffff8802ab3e2a80 CPU: 3 COMMAND: "txg_sync"
#0 [ffff8802ab3f3790] __schedule at ffffffff816cb514
#1 [ffff8802ab3f3840] schedule at ffffffff816cbc10
#2 [ffff8802ab3f3860] schedule_timeout at ffffffff816ce7ca
#3 [ffff8802ab3f3920] io_schedule_timeout at ffffffff816cb1f4
#4 [ffff8802ab3f3960] cv_wait_common at ffffffffa08dc072 [spl]
#5 [ffff8802ab3f39e0] __cv_wait_io at ffffffffa08dc228 [spl]
#6 [ffff8802ab3f39f0] zio_wait at ffffffffa09ee14f [zfs]
#7 [ffff8802ab3f3a40] dbuf_read at ffffffffa0945786 [zfs]
#8 [ffff8802ab3f3aa0] dmu_buf_hold at ffffffffa094db61 [zfs]
#9 [ffff8802ab3f3ae0] zap_lockdir at ffffffffa09b940a [zfs]
#10 [ffff8802ab3f3b80] zap_update at ffffffffa09ba653 [zfs]
#11 [ffff8802ab3f3bf0] spa_errlog_sync at ffffffffa099986f [zfs]
#12 [ffff8802ab3f3cc0] spa_sync at ffffffffa09916cf [zfs]
#13 [ffff8802ab3f3da0] txg_sync_thread at ffffffffa09a3dea [zfs]
#14 [ffff8802ab3f3e80] thread_generic_wrapper at ffffffffa08d5b11 [spl]
#15 [ffff8802ab3f3ec0] kthread at ffffffff810a263c
#16 [ffff8802ab3f3f50] ret_from_fork at ffffffff816cfacf
– snip –

The spa_errlog_lock will be dropped by spa_errlog_sync() only after the IOs complete.
The failmode property is set to "wait" for this pool and  spa_sync() is blocked waiting for
the IOs to complete.

Need to find a way to avoid blocking here and return failure and bailout.
Exploring it further.

**Comment by sanjeevbagewadi** [comment]
Also, "zpool clear" either :
- fails to clear (if failmode=continue) the pool although the underlying iSCSI device has recovered.
- And if failmode=wait, the zfs_ioc_clear() although it has failed to reopen the device, does not report
  an error. And hence, "zpool clear" wrongly assumes that the zfs_ioc_clear() succeeded and issues
  zfs_ioc_log_history() which gets stuck in the kernel at dmu_tx_assign() waiting for spa_sync() to
  complete.

-- snip --
PID: 21037 TASK: ffff8800ba96aa80 CPU: 2 COMMAND: "zpool"
#0 [ffff880220323b30] __schedule at ffffffff816cb514
#1 [ffff880220323be0] schedule at ffffffff816cbc10
#2 [ffff880220323c00] cv_wait_common at ffffffffa08d9275 [spl]
#3 [ffff880220323c80] __cv_wait at ffffffffa08d9305 [spl]
#4 [ffff880220323c90] txg_wait_synced at ffffffffa0999c99 [zfs]
#5 [ffff880220323ce0] dmu_tx_wait at ffffffffa095426e [zfs]
#6 [ffff880220323d50] dmu_tx_assign at ffffffffa095436a [zfs]
#7 [ffff880220323d80] spa_history_log_nvl at ffffffffa099106e [zfs]
#8 [ffff880220323dd0] spa_history_log at ffffffffa0991194 [zfs]
#9 [ffff880220323e00] zfs_ioc_log_history at ffffffffa09c380b [zfs]
#10 [ffff880220323e40] zfsdev_ioctl at ffffffffa09c4357 [zfs]
#11 [ffff880220323eb0] do_vfs_ioctl at ffffffff81216072
#12 [ffff880220323f00] sys_ioctl at ffffffff81216402
#13 [ffff880220323f50] entry_SYSCALL_64_fastpath at ffffffff816cf76e
RIP: 00007f17bcc9ba77 RSP: 00007ffd708a7f98 RFLAGS: 00000246
RAX: ffffffffffffffda RBX: 0000000000000000 RCX: 00007f17bcc9ba77
RDX: 00007ffd708a7fa0 RSI: 0000000000005a3f RDI: 0000000000000003
RBP: 0000000001157060 R8: 0000000001157f20 R9: 616332666162362d
R10: 0000000039373030 R11: 0000000000000246 R12: 000000000061fce0
R13: 000000000061e980 R14: 000000000061ead0 R15: 000000000000000e
ORIG_RAX: 0000000000000010 CS: 0033 SS: 002b
-- snip --

Upon further analysis found that the vdev_reopen() was failing because, the IOs issued
by vdev_probe() were failing.  vdev_accessible() called by zio_vdev_io_start() was failing
and this was because, vdev->vdev_remove_wanted was set on the vdev. This was set when the
vdev originally faulted. And this should have been cleared by spa_async_remove() which should have been spawned by spa_sync()->spa_async_dispatch(). However, the spa_sync() thread for this pool is stuck waiting for sync-IOs to complete.
-- snip --
PID: 24311 TASK: ffff8802a03d6a40 CPU: 2 COMMAND: "txg_sync"
#0 [ffff8802a03eb940] __schedule at ffffffff816cb514
#1 [ffff8802a03eb9f0] schedule at ffffffff816cbc10
#2 [ffff8802a03eba10] schedule_timeout at ffffffff816ce7ca
#3 [ffff8802a03ebad0] io_schedule_timeout at ffffffff816cb1f4
#4 [ffff8802a03ebb10] cv_wait_common at ffffffffa08d6072 [spl]
#5 [ffff8802a03ebb90] __cv_wait_io at ffffffffa08d6228 [spl]
#6 [ffff8802a03ebba0] zio_wait at ffffffffa0a43bdd [zfs]
#7 [ffff8802a03ebbf0] dsl_pool_sync_mos at ffffffffa09993e7 [zfs]
#8 [ffff8802a03ebc20] dsl_pool_sync at ffffffffa099a1e2 [zfs]
#9 [ffff8802a03ebcb0] spa_sync at ffffffffa09be347 [zfs]
#10 [ffff8802a03ebd60] txg_sync_thread at ffffffffa09d8681 [zfs]
#11 [ffff8802a03ebe80] thread_generic_wrapper at ffffffffa08cfb11 [spl]
#12 [ffff8802a03ebec0] kthread at ffffffff810a263c
#13 [ffff8802a03ebf50] ret_from_fork at ffffffff816cfacf
-- snip --

So, that is sort of a soft deadlock because, the zpool cannot be cleared unless the "vdev_remove_wanted" is handled and spa_sync() which is supposed to clear it, won't get there unless the IOs complete.

One possible fix is to clear the vdev_remove_wanted flag in vdev_clear().

**Comment by ikozhukhov** [comment]
probably similar to illumos https://www.illumos.org/issues/8593

**Comment by behlendorf** [comment]
Between the #5907 and #6399 fixes this should be resolved in 0.7.1.  If that's not the case, let's open a new issue to track the current problem.
